client/src/main/java/io/pravega/client/EventStreamClientFactory.java

```diff
@@ -0,0 +1,115 @@
+import io.pravega.client.stream.EventWriterConfig;
+import io.pravega.client.stream.Position;
+import io.pravega.client.stream.ReaderConfig;
+import io.pravega.client.stream.ReaderGroup;
+import io.pravega.client.stream.Serializer;
+import io.pravega.client.stream.TransactionalEventStreamWriter;
+import io.pravega.client.stream.impl.ClientFactoryImpl;
+import io.pravega.client.stream.impl.ControllerImpl;
+import io.pravega.client.stream.impl.ControllerImplConfig;
+import lombok.val;
+
+/**
+ * Used to create Writers and Readers operating on a stream.
+ * <p>
+ * Events that are written to a stream can be read by a reader. All events can be processed with
+ * exactly once semantics provided the reader has the ability to restore to the correct position
+ * upon failure. See {@link EventRead#getPosition()}
+ * <p>
+ * A note on ordering: Events inside of a stream have an order, but may need to be divided between
+ * multiple readers for scaling. In order to process events in parallel on different hosts and still
+ * have some ordering guarantees; events written to a stream have a routingKey see
+ * {@link EventStreamWriter#writeEvent(String, Object)}. Events with the same routing key are
+ * delivered to a reader in order serially (i.e. They must go the the same reader or its
+ * replacement). However because {@link ReaderGroup}s process events in parallel there is no
+ * ordering between events sent to different routing keys because these may end up on different
+ * readers.
+ *
+ * <p>
+ * A note on scaling: Because a stream can grow in its event rate, streams are divided into
+ * Segments. For the most part this is an implementation detail. However its worth understanding
+ * that the way a stream is divided between multiple readers in a group that wish to split the
+ * messages between them is by giving different segments to different readers.
+ */
+public interface EventStreamClientFactory extends AutoCloseable {
+
+    /**
+     * Creates a new instance of Client Factory.
+     *
+     * @param scope The scope string.
+     * @param config Configuration for the client.
+     * @return Instance of ClientFactory implementation.
+     */
+    static EventStreamClientFactory withScope(String scope, ClientConfig config) {
+        val connectionFactory = new ConnectionFactoryImpl(config);
+        return new ClientFactoryImpl(scope, new ControllerImpl(ControllerImplConfig.builder().clientConfig(config).build(),
+                connectionFactory.getInternalExecutor()), connectionFactory);
+    }
+
+    /**
+     * Creates a new writer that can write to the specified stream.
+     *
+     * @param streamName The name of the stream to write to.
+     * @param config The writer configuration.
+     * @param s The Serializer.
+     * @param <T> The type of events.
+     * @return Newly created writer object
+     */
+    <T> EventStreamWriter<T> createEventWriter(String streamName, Serializer<T> s, EventWriterConfig config);
```

Comment:
It is common to see instantiations of writers making use of `EventWriterConfig.builder().build()` as last argument (e.g., see [this sample](https://github.com/pravega/pravega-samples/blob/master/pravega-client-examples/src/main/java/io/pravega/example/gettingstarted/HelloWorldWriter.java)). Wondering if an overloaded version of this method would make sense with the only 2 first parameters. Maybe this idea may not be only limited to writers, but readers as well.

Reply:
I don't know if I want to remove the parameter. But I can see having a EventWriterConfig.DEFUALT or something like that. Either way, I don't want to add more changes into this PR.